tokio-quiche/src/socket/multi.rs

```diff
@@ -0,0 +1,94 @@
+use std::net::SocketAddr;
+use std::sync::Arc;
+use tokio::net::UdpSocket;
+
+use super::Socket;
+
+#[derive(Debug, Clone)]
+pub struct MultiSocket<Tx, Rx> {
+    pub paths: Arc<Vec<Socket<Tx, Rx>>>,
+    pub peer_addr: SocketAddr,
+}
+
+impl<Tx, Rx> MultiSocket<Tx, Rx> {
+    pub fn new(paths: Vec<Socket<Tx, Rx>>) -> std::io::Result<Self> {
+        if paths.is_empty() {
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::InvalidInput,
+                "Must provide at least one socket path",
+            ));
+        }
+
+        let first_peer = paths[0].peer_addr;
+        if !paths.iter().all(|p| p.peer_addr == first_peer) {
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::InvalidInput,
+                "All socket paths must have the same peer address",
+            ));
+        }
+
+        Ok(Self {
+            paths: Arc::new(paths),
+            peer_addr: first_peer,
+        })
+    }
+
+    pub fn primary_path(&self) -> Option<&Socket<Tx, Rx>> {
```

Comment:
Note that Path ID 0 may be closed for some reason; I assume the primary path is the "default" one?